File: src/shared/upstream.py
"""منبع واحد مقادیر متغیر upstream را از YAML می‌خواند."""

from functools import lru_cache
import os
from pathlib import Path

import yaml


@lru_cache(maxsize=1)
def upstream_config() -> dict:
    path = _config_path()
    data = yaml.safe_load(path.read_text(encoding="utf-8"))
    if not isinstance(data, dict) or data.get("version") != 1:
        raise ValueError(f"Invalid upstream config: {path}")
    return data


def upstream_value(section: str, name: str) -> str:
    value = upstream_config().get(section, {}).get(name)
    if not isinstance(value, str) or not value:
        raise ValueError(f"Missing upstream config value: {section}.{name}")
    return value


def _config_path() -> Path:
    explicit = os.getenv("AISTUDIO_UPSTREAM_CONFIG", "").strip()
    candidates = [
        Path(explicit) if explicit else None,
        Path("/app/config/upstream.yaml"),
        Path(__file__).resolve().parents[2] / "config" / "upstream.yaml",
    ]
    for candidate in candidates:
        if candidate and candidate.is_file():
            return candidate
    raise FileNotFoundError("config/upstream.yaml was not found")

File: internal/aistudio/upstream.go
package aistudio

import (
	"fmt"
	"os"
	"path/filepath"
	"sync"

	"gopkg.in/yaml.v3"
)

type Upstream struct {
	Version     int               `yaml:"version"`
	AIStudio    map[string]string `yaml:"aistudio"`
	Runtime     map[string]string `yaml:"runtime"`
	MakerSuite  map[string]string `yaml:"makersuite"`
	Drive       map[string]string `yaml:"drive"`
	Attestation map[string]string `yaml:"attestation"`
	Opaque      map[string]string `yaml:"opaque"`
}

var upstreamOnce sync.Once
var upstreamConfig Upstream
var upstreamError error

func LoadUpstream() (Upstream, error) {
	upstreamOnce.Do(func() {
		path := os.Getenv("AISTUDIO_UPSTREAM_CONFIG")
		if path == "" {
			for _, candidate := range []string{"/app/config/upstream.yaml", "config/upstream.yaml"} {
				if _, err := os.Stat(candidate); err == nil {
					path = candidate
					break
				}
			}
		}
		if path == "" {
			upstreamError = fmt.Errorf("config/upstream.yaml was not found")
			return
		}
		data, err := os.ReadFile(filepath.Clean(path))
		if err != nil {
			upstreamError = err
			return
		}
		if err := yaml.Unmarshal(data, &upstreamConfig); err != nil {
			upstreamError = err
			return
		}
		if upstreamConfig.Version != 1 {
			upstreamError = fmt.Errorf("invalid upstream config: %s", path)
		}
	})
	return upstreamConfig, upstreamError
}

func mustValue(section map[string]string, name string) string {
	value := section[name]
	if value == "" {
		panic("missing upstream config value: " + name)
	}
	return value
}
